File: golang/gray_code.go
package main

// Space : O(2^n)
// Time  : O(2^n)

func grayCode(n int) []int {
    ans := []int{0}
    
    for i := 1; i <= n; i++ {
        prev_len := len(ans)
        mask := 1 << (i-1)
        for j := prev_len; j > 0; j-- {
            ans = append(ans, mask + ans[j-1])
        }
    }
    return ans
}

File: python3/gray_code.py
# Space : O(2^n)
# Time  : O(2^n)

class Solution:
    def grayCode(self, n: int) -> List[int]:
        ans = []
        ans.append(0)

        for i in range(1, n+1):
            prev_len = len(ans)
            mask = 1 << (i-1)
            for j in range(prev_len, 0, -1):
                ans.append(mask + ans[j-1])
        return ans
